File: ports/go/internal/svc/store/store.go
// Package store is the storage contract and its filesystem implementation.
//
// The service has no database on purpose — the data is a scratch pad that dies with the
// container. But "no database" must not mean "no abstraction", or moving to SQL later
// becomes a rewrite. So this package presents the surface a database session would, and the
// repositories on top of it take the store as their first argument, exactly as the
// reference's repository functions take `db`.
//
// **SQL SWAP POINT.** Implementing DocumentStore over a real database, and constructing
// that instead of FileStore, is the whole migration as far as callers are concerned.
// Router and worker code does not change.
//
// On-disk layout:
//
//	$DATA_DIR/
//	  documents/42/
//	    record.json     the "row"
//	    original.jpg    exactly the bytes uploaded
//	    canvas.png      the deskewed/rectified canvas
//	    result.json     the full recognition view model
//	  api_keys.json
//	  settings.json
//	  users.json      named accounts (AUTH_MODE=users), Argon2id PHC hashes only
//	  audit.jsonl     the action log, one JSON object per line, last AuditMaxEntries kept
//
// Four design notes worth reading before changing anything here:
//
//   - **The index lives in memory; disk is scanned once at startup.** The service is
//     pinned to ONE process — the pipeline singleton and this index both are — so a
//     shared in-memory index is legitimate rather than a shortcut.
//   - **Writes are atomic** (temp file plus rename, atomic on NTFS and ext4). A
//     half-written record.json would survive a crash and poison the next boot.
//   - **Reads return COPIES.** A live shared instance would let one request's edit leak
//     into another's view, so Update returns a NEW record and callers must rebind. That
//     is already how the reference's routers are written (`m = repo.update_status(...)`),
//     so the idiom ports without changing call sites.
//   - **`result` is not held in the index** — it can be 100 KB of boxes per document.
//     GetByID loads it lazily; list queries never touch it.
//
// Port of service/core/store.py and service/core/database.py.
package store

import (
	"github.com/protei300/RussianDocsOCR/ports/go/internal/svc/model"
)

// SortColumns is the whitelist of sortable columns, shared by every backend so they cannot
// drift apart.
//
// A whitelist rather than dynamic field lookup: in a SQL backend that difference is an
// injection vector, and here it is what stops a typo in a query string from silently
// sorting by nothing.
var SortColumns = map[string]bool{
	"created_at": true, "filename": true, "status": true, "doc_type": true,
	"doc_conf": true, "processing_ms": true, "size_bytes": true,
}

// Query is the filter/sort/page request for a document listing.
//
// A struct rather than a long parameter list because it crosses the store boundary and
// grows: a positional call is where "swap date_from and date_to" hides.
type Query struct {
	Status   string
	DocType  string
	Search   string
	DateFrom string
	DateTo   string
	Page     int
	PageSize int
	SortBy   string
	SortDir  string
}

// Stats is the aggregate summary the status page shows.
type Stats struct {
	Queued          int  `json:"queued"`
	Processing      int  `json:"processing"`
	Done            int  `json:"done"`
	Failed          int  `json:"failed"`
	Total           int  `json:"total"`
	Recognised      int  `json:"recognised"`
	AvgProcessingMs *int `json:"avg_processing_ms"`
}

// DocumentStore is everything the service needs from a storage backend.
//
// Query methods live HERE rather than in the repository functions, deliberately: filtering
// a list in memory is correct for a few hundred JSON files and wrong for a table, so each
// backend has to express "the newest twenty matching rows" in its own terms. Putting the
// queries behind the interface is what lets it.
type DocumentStore interface {
	// Backend is "files" or "sql" — surfaced on the status page, because "why did my
	// data vanish" is answered by this one word.
	Backend() string
	// IsEphemeral reports whether the contents survive a restart.
	IsEphemeral() bool

	NextDocumentID() int
	GetRecord(id int) *model.Document
	PutRecord(rec *model.Document) *model.Document
	DropRecord(id int)
	QueryDocuments(q Query) ([]*model.Document, int)
	AllRecords() []*model.Document
	NextQueuedID() (int, bool)
	QueuePosition(id int) (int, bool)
	CountByStatus() map[string]int
	AggregateStats() Stats

	SaveResultPayload(id int, payload map[string]any) error
	LoadResultPayload(id int) map[string]any

	AllApiKeys() []*model.ApiKey
	NextApiKeyID() int
	PutApiKey(key *model.ApiKey) (*model.ApiKey, error)
	DropApiKey(id int) (bool, error)

	AllSettings() map[string]string
	SetSettings(values map[string]string) (map[string]string, error)

	// Accounts. Present in both auth modes, but only ever populated in AUTH_MODE=users. Every
	// read returns a COPY: the repository mutates what it gets back, and handing out the
	// indexed instance made the "last administrator" rule a check-then-act race against
	// whoever else held the same object (the first Python version did exactly that).
	AllUsers() []*model.User // ordered by id
	GetUser(id int) *model.User
	// FindUser matches the username case-insensitively, after trimming: an account created
	// as "Admin" that cannot be used by typing "admin" is a support ticket, and two accounts
	// differing only in case are an impersonation waiting to happen.
	FindUser(username string) *model.User
	// NextUserID does NOT reserve the id; PutUser advances the counter. Callers allocate and
	// store under the repository's write lock, which is what makes that safe.
	NextUserID() int
	PutUser(user *model.User) (*model.User, error)
	DropUser(id int) (bool, error)

	// The action log. AppendAudit assigns the id and never fails the caller — see FileStore.
	AppendAudit(entry model.AuditEntry) model.AuditEntry
	// RecentAudit is newest first: action is an exact match, actor a case-insensitive
	// substring, either "" for no filter.
	RecentAudit(limit int, action, actor string) []model.AuditEntry

	// DocDir is a plain directory in every backend: binary artifacts stay on the
	// filesystem regardless of where the metadata lives.
	DocDir(id int) string
	DiskUsageBytes() int64
}

File: service/core/store.py
"""The storage contract both backends implement.

Two backends exist:

* ``FileStore`` (``core.database``) — JSON files on disk, wiped at every start.
  The zero-configuration default so the service runs out of the box.
* ``SqlStore`` (``core.db_sql``) — any SQLAlchemy-supported database. What you
  use when the data matters.

Everything above this layer — routers, worker, repositories — is written against
the protocol below and cannot tell them apart. That is what makes swapping the
backend a configuration change rather than a rewrite, and
``tests/service/test_repository_contract.py`` is what keeps the two honest by
running the same suite against both.

Query methods live *here*, not in the repository functions, on purpose. Filtering
a list in Python is correct for a few hundred JSON files and wrong for a table:
each backend needs to express "the newest twenty matching rows" in its own
terms. Putting the queries behind the protocol lets it.
"""
from __future__ import annotations

from pathlib import Path
from typing import Any, Protocol, runtime_checkable

from service.core.models import ApiKey, DocumentRecord


@runtime_checkable
class DocumentStore(Protocol):
    """Everything the service needs from a storage backend."""

    # -- identity ------------------------------------------------------------
    @property
    def backend(self) -> str:
        """``'files'`` or ``'sql'`` — surfaced on the status page."""

    @property
    def is_ephemeral(self) -> bool:
        """True when the contents do not survive a restart."""

    # -- documents -----------------------------------------------------------
    def next_document_id(self) -> int: ...

    def get_record(self, doc_id: int) -> DocumentRecord | None: ...

    def put_record(self, record: DocumentRecord) -> DocumentRecord: ...

    def drop_record(self, doc_id: int) -> None: ...

    def query_documents(self, *, status: str | None = None, doc_type: str | None = None,
                        search: str | None = None, date_from: str | None = None,
                        date_to: str | None = None, page: int = 1, page_size: int = 20,
                        sort_by: str = "created_at",
                        sort_dir: str = "desc") -> tuple[list[DocumentRecord], int]:
        """One page of matching records, plus the unpaged total."""

    def all_records(self) -> list[DocumentRecord]:
        """Every record. Only for small-scale operations (purge, recovery)."""

    def next_queued_id(self) -> int | None: ...

    def queue_position(self, doc_id: int) -> int | None: ...

    def count_by_status(self) -> dict[str, int]: ...

    def aggregate_stats(self) -> dict[str, Any]: ...

    # -- results -------------------------------------------------------------
    def save_result_payload(self, doc_id: int, payload: dict[str, Any]) -> None:
        """Persist the recognition view model for a document."""

    def load_result_payload(self, doc_id: int) -> dict[str, Any] | None: ...

    # -- api keys ------------------------------------------------------------
    def all_api_keys(self) -> list[ApiKey]: ...

    def next_api_key_id(self) -> int: ...

    def put_api_key(self, key: ApiKey) -> ApiKey: ...

    def drop_api_key(self, key_id: int) -> bool: ...

    # -- settings ------------------------------------------------------------
    def all_settings(self) -> dict[str, str]: ...

    def set_settings(self, values: dict[str, str]) -> dict[str, str]: ...

    # -- artifacts -----------------------------------------------------------
    # Binary files stay on the filesystem regardless of backend, so this is a
    # plain directory in both cases.
    def doc_dir(self, doc_id: int) -> Path: ...

    def disk_usage_bytes(self) -> int: ...


#: Sortable columns, shared by both backends so they cannot drift apart.
#: A whitelist rather than dynamic attribute access — in the SQL backend that
#: difference is an injection vector.
SORT_COLUMNS = frozenset({
    "created_at", "filename", "status", "doc_type", "doc_conf",
    "processing_ms", "size_bytes",
})
